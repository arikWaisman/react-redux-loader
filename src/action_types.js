export const INTERFACE_REQUEST = 'react-redux-modal/INTERFACE_REQUEST';
export const INTERFACE_COMPLETE = 'react-redux-modal/INTERFACE_COMPLETE';
export const INTERFACE_ERROR = 'react-redux-modal/INTERFACE_ERROR';

export const SHOW_LOADER = 'react-redux-modal/SHOW_LOADER';
export const HIDE_LOADER = 'react-redux-modal/HIDE_LOADER';

export const SHOW_INLINE_LOADER = 'react-redux-modal/SHOW_INLINE_LOADER';
export const HIDE_INLINE_LOADER = 'react-redux-modal/HIDE_INLINE_LOADER';
