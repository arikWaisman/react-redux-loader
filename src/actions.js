import {
  SHOW_LOADER,
  HIDE_LOADER,
  SHOW_INLINE_LOADER,
  HIDE_INLINE_LOADER,
  INTERFACE_REQUEST,
  INTERFACE_COMPLETE,
} from './action_types';

export const showSyncLoader = (args = {}) => {
  const { type = null, shouldTimeOut = null, timeOut = null } = args;
  return {
    type: SHOW_LOADER,
    payload: {
      type,
      ...(shouldTimeOut ? { shouldWaitToDisplay: true, timeOut } : {}),
    },
  };
};

export const hideSyncLoader = () => ({
  type: HIDE_LOADER,
});

export const showInlineLoader = (key = null) => ({
  type: SHOW_INLINE_LOADER,
  payload: {
    key,
  },
});

export const hideInlineLoader = (key = null) => ({
  type: HIDE_INLINE_LOADER,
  payload: {
    key,
  },
});

export const interfaceRequest = (args = {}) => {
  const { type = null, shouldTimeOut = null, timeOut = null } = args;
  return {
    type: INTERFACE_REQUEST,
    payload: {
      type,
      ...(shouldTimeOut ? { shouldWaitToDisplay: true, timeOut } : {}),
      async: true,
    },
  };
};

export const interfaceComplete = () => ({
  type: INTERFACE_COMPLETE,
});
