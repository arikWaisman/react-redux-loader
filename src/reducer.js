import {
  INTERFACE_REQUEST,
  INTERFACE_COMPLETE,
  INTERFACE_ERROR,
  SHOW_LOADER,
  HIDE_LOADER,
  SHOW_INLINE_LOADER,
  HIDE_INLINE_LOADER,
} from './action_types';

const INITIAL_STATE = {
  requests: 0,
  showLoader: null,
  error: null,
  loaderType: null,
  shouldWaitToDisplay: false,
  timeOut: null,
  async: false,
  inline: [],
};

// keep loader type if already one open
const getLoaderType = (requestedType = null, state) =>
  state.requests > 0 ? state.loaderType : requestedType;

// remove an inline loader from inline list
const updateInlineLoaders = (loaderToRemove, state) => {
  const indexToRemove = state.inline.indexOf(loaderToRemove);

  return state.inline.filter((key, index) => index !== indexToRemove);
};

export default function(state = INITIAL_STATE, action) {
  switch (action.type) {
    case INTERFACE_REQUEST:
      return {
        ...state,
        requests: state.requests + 1,
        loaderType: getLoaderType(action.payload.type, state),
        shouldWaitToDisplay:
          action.payload.shouldWaitToDisplay || state.shouldWaitToDisplay,
        timeOut: action.payload.timeOut || state.timeOut,
        async: true,
      };

    case INTERFACE_COMPLETE:
      return {
        ...state,
        requests: state.requests - 1,
        ...(state.requests - 1 > 0
          ? {}
          : {
              shouldWaitToDisplay: false,
              timeOut: null,
              async: false,
              showLoader: false,
            }),
      };

    case SHOW_LOADER:
      return {
        ...state,
        showLoader: true,
        loaderType: action.payload.type,
        shouldWaitToDisplay:
          action.payload.shouldWaitToDisplay || state.shouldWaitToDisplay,
        timeOut: action.payload.timeOut || state.timeOut,
      };

    case HIDE_LOADER:
      return { ...state, showLoader: false, async: false, loaderType: null };

    case SHOW_INLINE_LOADER:
      return { ...state, inline: [...state.inline, action.payload.key] };

    case HIDE_INLINE_LOADER:
      return {
        ...state,
        inline: updateInlineLoaders(action.payload.key, state),
      };

    case INTERFACE_ERROR:
      return { ...state, error: action.error };

    default:
      return state;
  }
}
