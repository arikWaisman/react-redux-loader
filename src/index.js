import LoaderContext from './loader_context';
import * as actions from './actions';
import * as actionTypes from './action_types';
import loaderReducer from './reducer';
import makeLoaderRoot from './components/loader_root';
import LoaderFactory from './components/loader_factory';

export {
  actions,
  actionTypes,
  makeLoaderRoot,
  loaderReducer,
  LoaderFactory,
  LoaderContext,
};
