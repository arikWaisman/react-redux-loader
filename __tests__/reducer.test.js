import { loaderReducer, actionTypes } from '../src';

describe('Loader reducer', () => {
  describe('default case with initial state', () => {
    test('return default case', () => {
      const action = { type: 'FAKE_ACTION' };

      expect(loaderReducer(undefined, action)).toMatchSnapshot();
    });
  });

  describe('show sync loader', () => {
    test('returns the correct state', () => {
      const action = {
        type: actionTypes.SHOW_LOADER,
        payload: { type: 'FAKE_LOADER' },
      };

      expect(loaderReducer(undefined, action)).toMatchSnapshot();
    });
  });

  describe('hide sync loader', () => {
    test('returns the correct state', () => {
      const action = {
        type: actionTypes.HIDE_LOADER,
      };

      expect(loaderReducer(undefined, action)).toMatchSnapshot();
    });
  });

  describe('show and then hide sync loader', () => {
    test('returns the correct state', () => {
      const showAction = {
        type: actionTypes.SHOW_LOADER,
        payload: { type: 'FAKE_LOADER' },
      };

      const hideAction = {
        type: actionTypes.HIDE_LOADER,
      };

      const stateAfterShow = loaderReducer(undefined, showAction);
      expect(loaderReducer(stateAfterShow, hideAction)).toMatchSnapshot();
    });
  });

  describe('show inline loader', () => {
    test('returns the correct state', () => {
      const action = {
        type: actionTypes.SHOW_INLINE_LOADER,
        payload: { key: 'FAKE_INLINE_LOADER_KEY' },
      };

      expect(loaderReducer(undefined, action)).toMatchSnapshot();
    });
  });

  describe('hide inline loader', () => {
    test('returns the correct state', () => {
      const action = {
        type: actionTypes.HIDE_INLINE_LOADER,
        payload: { key: 'FAKE_INLINE_LOADER_KEY' },
      };

      expect(loaderReducer(undefined, action)).toMatchSnapshot();
    });
  });

  describe('show and then hide inline loader', () => {
    test('returns the correct state', () => {
      const showAction = {
        type: actionTypes.SHOW_INLINE_LOADER,
        payload: { key: 'FAKE_INLINE_LOADER_KEY' },
      };

      const hideAction = {
        type: actionTypes.HIDE_INLINE_LOADER,
        payload: { key: 'FAKE_INLINE_LOADER_KEY' },
      };

      const stateAfterShow = loaderReducer(undefined, showAction);
      expect(loaderReducer(stateAfterShow, hideAction)).toMatchSnapshot();
    });
  });

  describe('show async loader with an interface request', () => {
    test('returns the correct state', () => {
      const action = {
        type: actionTypes.INTERFACE_REQUEST,
        payload: { type: 'FAKE_LOADER' },
      };

      expect(loaderReducer(undefined, action)).toMatchSnapshot();
    });
  });

  describe('interface complete returns proper state', () => {
    test('returns the correct state', () => {
      const action = {
        type: actionTypes.INTERFACE_COMPLETE,
      };

      expect(loaderReducer(undefined, action)).toMatchSnapshot();
    });
  });

  describe('interface error returns proper state', () => {
    test('returns the correct state', () => {
      const action = {
        type: actionTypes.INTERFACE_ERROR,
        error: 'some fake error',
      };

      expect(loaderReducer(undefined, action)).toMatchSnapshot();
    });
  });

  describe('show async loader with more than one request and ensure complete decrements ', () => {
    test('returns the correct state', () => {
      const showAction = {
        type: actionTypes.INTERFACE_REQUEST,
        payload: { type: 'FAKE_LOADER' },
      };

      const hideAction = {
        type: actionTypes.INTERFACE_COMPLETE,
        payload: { key: 'FAKE_INLINE_LOADER_KEY' },
      };

      const stateAfterFirstShow = loaderReducer(undefined, showAction);
      const stateAfterSecondShow = loaderReducer(
        stateAfterFirstShow,
        showAction,
      );

      expect(loaderReducer(stateAfterSecondShow, hideAction)).toMatchSnapshot();
    });
  });
});
