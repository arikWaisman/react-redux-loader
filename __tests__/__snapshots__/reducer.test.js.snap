// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Loader reducer default case with initial state return default case 1`] = `
Object {
  "async": false,
  "error": null,
  "inline": Array [],
  "loaderType": null,
  "requests": 0,
  "shouldWaitToDisplay": false,
  "showLoader": null,
  "timeOut": null,
}
`;

exports[`Loader reducer hide inline loader returns the correct state 1`] = `
Object {
  "async": false,
  "error": null,
  "inline": Array [],
  "loaderType": null,
  "requests": 0,
  "shouldWaitToDisplay": false,
  "showLoader": null,
  "timeOut": null,
}
`;

exports[`Loader reducer hide sync loader returns the correct state 1`] = `
Object {
  "async": false,
  "error": null,
  "inline": Array [],
  "loaderType": null,
  "requests": 0,
  "shouldWaitToDisplay": false,
  "showLoader": false,
  "timeOut": null,
}
`;

exports[`Loader reducer interface complete returns proper state returns the correct state 1`] = `
Object {
  "async": false,
  "error": null,
  "inline": Array [],
  "loaderType": null,
  "requests": -1,
  "shouldWaitToDisplay": false,
  "showLoader": false,
  "timeOut": null,
}
`;

exports[`Loader reducer interface error returns proper state returns the correct state 1`] = `
Object {
  "async": false,
  "error": "some fake error",
  "inline": Array [],
  "loaderType": null,
  "requests": 0,
  "shouldWaitToDisplay": false,
  "showLoader": null,
  "timeOut": null,
}
`;

exports[`Loader reducer show and then hide inline loader returns the correct state 1`] = `
Object {
  "async": false,
  "error": null,
  "inline": Array [],
  "loaderType": null,
  "requests": 0,
  "shouldWaitToDisplay": false,
  "showLoader": null,
  "timeOut": null,
}
`;

exports[`Loader reducer show and then hide sync loader returns the correct state 1`] = `
Object {
  "async": false,
  "error": null,
  "inline": Array [],
  "loaderType": null,
  "requests": 0,
  "shouldWaitToDisplay": false,
  "showLoader": false,
  "timeOut": null,
}
`;

exports[`Loader reducer show async loader with an interface request returns the correct state 1`] = `
Object {
  "async": true,
  "error": null,
  "inline": Array [],
  "loaderType": "FAKE_LOADER",
  "requests": 1,
  "shouldWaitToDisplay": false,
  "showLoader": null,
  "timeOut": null,
}
`;

exports[`Loader reducer show async loader with more than one request and ensure complete decrements  returns the correct state 1`] = `
Object {
  "async": true,
  "error": null,
  "inline": Array [],
  "loaderType": "FAKE_LOADER",
  "requests": 1,
  "shouldWaitToDisplay": false,
  "showLoader": null,
  "timeOut": null,
}
`;

exports[`Loader reducer show inline loader returns the correct state 1`] = `
Object {
  "async": false,
  "error": null,
  "inline": Array [
    "FAKE_INLINE_LOADER_KEY",
  ],
  "loaderType": null,
  "requests": 0,
  "shouldWaitToDisplay": false,
  "showLoader": null,
  "timeOut": null,
}
`;

exports[`Loader reducer show sync loader returns the correct state 1`] = `
Object {
  "async": false,
  "error": null,
  "inline": Array [],
  "loaderType": "FAKE_LOADER",
  "requests": 0,
  "shouldWaitToDisplay": false,
  "showLoader": true,
  "timeOut": null,
}
`;
